### DISCUSSION TITLE
Building strongswan with non-root privilege's not working

### DISCUSSION THREAD

**Discussion by davehouser1** [strongswan/strongswan#737] [Q&A]
I want to be able to install strongswan so that a specific non-root user is able to run swanctl, and possibly make changes to the configuration files (However the instructions for compiling strongswan for a non-root user does not detail if this is possible or not)

I read through these instructions [here](https://wiki.strongswan.org/projects/strongswan/wiki/ReducedPrivileges).

**Problem:** 
After make install completes I check the swanctl binary and its still owned by root:root, also not not able to edit strongswan config files as non-root user, they are owned by root:root as well with write permissions only to the owner.

This is the process I followed:

1.  Downloaded the strongswan package 
```
# wget https://download.strongswan.org/strongswan-5.9.4.tar.bz2
# tar -xvf strongswan-5.9.4.tar.bz2
# cd strongswan-5.9.4
```
2. I performed a configure with the specific options detailed in the instructions in the link above, then performed a make and make install (Note: all the following commands were run as root):
```
# ./configure --disable-static \
                      --sysconfdir=/etc/strongswan \
                      --with-ipsecdir=/usr/libexec/strongswan \
                      --bindir=/usr/libexec/strongswan \
                      --with-ipseclibdir=/usr/lib64/strongswan \
                      --with-piddir=/run/strongswan \
                      --with-nm-ca-dir=/etc/strongswan/ipsec.d/cacerts/ \
                      --enable-kernel-libipsec \
                      --with-user=user1 \
                      --with-group=user1
# make
# make install
```
3. I checked the configuration files and the swanctl binary, they both show being owned by root and execution privilege's by owner. I also noticed that this binary does not have any capabilities set to it either. I also checked the files in /etc/strongswan, they are all owned by root as well. "user1" is definitely created on the system and is listed in /etc/passwd.

```
#cat /etc/passwd | grep user1
user1:x:1003:1003::/home/user1:/bin/bash
```

Why is the build / install not setting the correct permissions on install? I can go around and adjust group ownership with chown , as well as chmod the execution on all related files, and set capabilities on binaries, but I thought doing all that manual work defeats the purpose of setting the configure parameters in the configuration during compiling, does it not?
What am I missing here?

**Comment by tobiasbrunner** [comment]
> I check the swanctl binary and its still owned by root:root

Which should not prevent regular users from running it, so that's irrelevant. What decides if a regular user can communicate via VICI (using `swanctl` or some other client) is whether the user/group has access to the UNIX socket that's created at runtime by the _vici_ plugin (before dropping privileges/switching user/group). You could also switch to TCP, then any user could communicate with the daemon, even from a different host if you don't put firewall rules in place or pin the socket to `localhost` to prevent it.

> also not not able to edit strongswan config files as non-root user,

You have to change their permission/owner manually (the configured user/group don't influence the `install` calls as that doesn't work well with packaging/cross-compilation, also note that the user/group of the daemon can be changed later via strongswan.conf). Alternatively, call `swanctl` with `--file` argument or `SWANCTL_DIR` env variable (if the latter is an empty string, the current working dir is used) to use files in a different location.

**Reply by davehouser1** [reply]
Thanks for the details. I understand now.
Once the systemd service is started, then I can run swanctl commands unprivileged user as its able to connect to the unix socket like you said. I also see in the journalctl output `dropped capabilities, running as uid 1003, gid 1003`.

At this point I am looking for details on how to set up a systemd service for the actual user to run.
I already have a systemctl --user service built for the user, and tested this with the user's dbus domain with a test service. So I know the user is able to run services with systemctl --user.

I tried copying over the systemd unit into that users .config directory (~user1/.config/systemd/user/strongswan.service)
I then adjusted the unit for `WantedBy=default.target`.

When trying to run the service with `systemctl --user start strongswan.service` I see the following errors in the journalctl output.
Is running the strongswan as a systemctl --user unit supported? 

```
Nov 08 22:15:04 gst-01b systemd[876]: Starting strongSwan IPsec IKEv1/IKEv2 daemon using swanctl...
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: PKCS11 module '<name>' lacks library path
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: TPM 2.0 - could not load "libtss2-tcti-tabrmd.so.0"
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: plugin 'tpm': failed to load - tpm_plugin_create returned NULL
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: openssl FIPS mode(0) - disabled
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: kernel-netlink plugin might require CAP_NET_ADMIN capability
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: farp plugin requires CAP_NET_RAW capability
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: plugin 'farp': failed to load - farp_plugin_create returned NULL
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: xauth-pam plugin requires CAP_AUDIT_WRITE capability
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: plugin 'xauth-pam': failed to load - xauth_pam_plugin_create returned NULL
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: dhcp plugin requires CAP_NET_BIND_SERVICE capability
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: plugin 'dhcp': failed to load - dhcp_plugin_create returned NULL
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: binding socket 'unix:///run/strongswan/charon.dck' failed: Address already in use
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: creating duplicheck socket failed
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: plugin 'duplicheck': failed to load - duplicheck_plugin_create returned NULL
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: getting SPD hash threshold failed: Operation not permitted (1)
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: getting SPD hash threshold failed: Operation not permitted (1)
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: unable to bind XFRM event socket: Operation not permitted (1)
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: socket-default plugin requires CAP_NET_BIND_SERVICE capability
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: feature CUSTOM:libcharon in critical plugin 'charon-systemd' has unmet dependency: CUSTOM:kernel-ipsec
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: received netlink error: Operation not permitted (1)
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: unable to create IPv4 routing table rule
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: received netlink error: Operation not permitted (1)
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: unable to create IPv6 routing table rule
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: loading ca certificates from '/etc/strongswan/ipsec.d/cacerts'
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: opening directory '/etc/strongswan/ipsec.d/cacerts' failed: Permission denied
Nov 08 22:15:04 gst-01b charon-systemd[2234577]:   reading directory failed
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: loading aa certificates from '/etc/strongswan/ipsec.d/aacerts'
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: opening directory '/etc/strongswan/ipsec.d/aacerts' failed: Permission denied
Nov 08 22:15:04 gst-01b charon-systemd[2234577]:   reading directory failed
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: loading ocsp signer certificates from '/etc/strongswan/ipsec.d/ocspcerts'
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: opening directory '/etc/strongswan/ipsec.d/ocspcerts' failed: Permission denied
Nov 08 22:15:04 gst-01b charon-systemd[2234577]:   reading directory failed
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: loading attribute certificates from '/etc/strongswan/ipsec.d/acerts'
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: opening directory '/etc/strongswan/ipsec.d/acerts' failed: Permission denied
Nov 08 22:15:04 gst-01b charon-systemd[2234577]:   reading directory failed
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: loading crls from '/etc/strongswan/ipsec.d/crls'
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: opening directory '/etc/strongswan/ipsec.d/crls' failed: Permission denied
Nov 08 22:15:04 gst-01b charon-systemd[2234577]:   reading directory failed
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: loading secrets from '/etc/strongswan/ipsec.secrets'
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: opening secrets file '/etc/strongswan/ipsec.secrets' failed: Permission denied
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: binding socket 'unix:///run/strongswan/charon.ctl' failed: Address already in use
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: creating stroke socket failed
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: binding socket 'unix:///run/strongswan/charon.vici' failed: Address already in use
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: creating vici socket failed
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: opening triplet file /etc/strongswan/ipsec.d/triplets.dat failed: Permission denied
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: loaded 0 RADIUS server configurations
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: HA config misses local/remote address
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: no script for ext-auth script defined, disabled
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: failed to load 1 critical plugin feature
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: received netlink error: Operation not permitted (1)
Nov 08 22:15:04 gst-01b charon-systemd[2234577]: received netlink error: Operation not permitted (1)
Nov 08 22:15:04 gst-01b systemd[876]: strongswan.service: Main process exited, code=exited, status=66/NOINPUT
Nov 08 22:15:04 gst-01b systemd[876]: strongswan.service: Failed with result 'exit-code'.
Nov 08 22:15:04 gst-01b systemd[876]: Failed to start strongSwan IPsec IKEv1/IKEv2 daemon using swanctl.
```

**Reply by tobiasbrunner** [reply]
You can't **start** the daemon as non-root user unless you give the executable the capabilities it requires (and e.g. change paths and or groups so it is able to create/access the files it requires). Also note that there are plugins loaded that you haven't actually built (at least not based on the configure options you've shown above), so make sure you remove files from previous builds (or binary packages), including config files you haven't modified.

**Reply by V0174** [reply]
Hi,
I found this discussion because I'm trying to run Strongswan as a non-root user as well.

You wrote four years ago:
> You can't **start** the daemon as non-root user unless you give the executable the capabilities it requires (and e.g. change paths and or groups so it is able to create/access the files it requires).

Page [Running strongSwan with Reduced Privileges](https://docs.strongswan.org/docs/latest/install/reducedPrivileges.html) now states:
> strongSwan allows to run its daemon under a non-root user.

Did something change in Strongswan? I understand that to _run_ strongswan as another user, one has to _start_ it as this user. I followed the instructions, but when I try to run the service with non-root privileges, I get:

```
Dec 18 17:01:17 server ipsec_starter[54241]: Starting strongSwan 6.0.4 IPsec [starter]...
Dec 18 17:01:17 server  ipsec_starter[54241]: permission denied (must be superuser)
Dec 18 17:01:17 server  systemd[1]: strongswan.service: Main process exited, code=exited, status=4/NOPERMISSION
```

**Reply by tobiasbrunner** [reply]
Nothing has changed. You either have to start it as root, then the daemon switches to a configured user. Or you assign the required capabilities to the daemon's executable and then start as non-root user. But note that you use the legacy service unit that doesn't directly start the IKE daemon but the old starter daemon, which checks the UID unless it was explicitly compiled with `STARTER_ALLOW_NON_ROOT`.

**Reply by V0174** [reply]
> compiled with `STARTER_ALLOW_NON_ROOT`

Thanks, that moved me forward. I understand why this is not mentioned on the page about non-root users as the ipsec_starter is now obsolete. Unfortunately we have to use it for the time being.

For anyone having the same issue, I have added this parameter to ./configure:
`./configure CFLAGS="-DSTARTER_ALLOW_NON_ROOT"`

It still didn't work though, the sockets could not be created:
```
Dec 19 09:22:41 00[NET] binding socket 'unix:///var/run/charon.ctl' failed: Address already in use
Dec 19 09:22:41 00[CFG] creating stroke socket failed
Dec 19 09:22:41 00[LIB] feature CUSTOM:stroke in plugin 'stroke' failed to load
Dec 19 09:22:41 00[NET] cannot change ownership of socket 'unix:///var/run/charon.vici' without CAP_CHOWN capability. socket directory should be accessible to UID/GID under which the daemon will run
Dec 19 09:22:41 00[NET] binding socket 'unix:///var/run/charon.vici' failed: Address already in use
Dec 19 09:22:41 00[CFG] creating vici socket failed
Dec 19 09:22:41 00[LIB] feature CUSTOM:vici in plugin 'vici' failed to load 
```
That makes sense. I would need to set the capabilities for the user/binary (or at least that was what AI proposed) as it needs to open a low UDP port. I followed this path for a while, but then it got too complicated so I settled for the "start as root, drop privileges later" approach.

@tobiasbrunner, thanks again for your quick reply and for your work on Strongswan!
